pkg/recordbot/recorder.go

```diff
@@ -0,0 +1,159 @@
+package recordbot
+
+import (
+	"errors"
+	"log"
+	"sync"
+
+	"github.com/pion/webrtc/v3"
+	"github.com/pion/webrtc/v3/pkg/media"
+	"github.com/pion/webrtc/v3/pkg/media/h264writer"
+	"github.com/pion/webrtc/v3/pkg/media/ivfwriter"
+	"github.com/pion/webrtc/v3/pkg/media/oggwriter"
+)
+
+var ErrUnsupportedCodec = errors.New("unsupported codec")
+
+type Recorder interface {
+	// Reads RTP packets in TrackRemote and save them to local file.
+	// No need to call this in a goroutine as the function already calls one.
+	Start(track *webrtc.TrackRemote)
+
+	// Stops the recording and does cleanups such as closing the file and channel.
+	Stop()
+}
+
+type recorder struct {
+	// Save file name to be passed into hooks
+	filename string
+
+	// Each recorder uses different writer depending if it's for video / audio
+	writer media.Writer
+
+	// The `done` channel is an internal channel for tracking the start to finish of the recording
+	done chan bool
+
+	// WaitGroup is used to ensure everything is cleaned up in the goroutines before closing the recorder
+	wg sync.WaitGroup
+
+	// Hooks provide access to customising aspects of the recorder, such as uploading the file after recording
+	hooks RecorderHooks
+}
+
+type RecorderHooks struct {
+	// The recorder will pass the filename that it was instantiated with, to the hook
+	UploadFile func(filename string) error
+}
+
+func NewRecorder(filename string, codec webrtc.RTPCodecParameters, hooks RecorderHooks) (Recorder, error) {
```

Comment:
sounds like the idea for this recorder is a single track? should we rename the type/constructor to indicate as such? maybe `NewSingleTrackRecorder`

it'd be great to one day add a recorder that would mux audio and video together into MP4

Reply:
That’s a good idea which I didn’t think of. Thanks for the suggestion! I’ll probably need to define an interface for the 2 recorders